src/runtime/Converter.cs

```diff
@@ -880,12 +881,49 @@ private static void SetConversionError(BorrowedReference value, Type target)
         /// Convert a Python value to a correctly typed managed array instance.
         /// The Python value must support the Python iterator protocol or and the
         /// items in the sequence must be convertible to the target array type.
+        /// If the Python value supports the buffer protocol the underlying buffer
+        /// will be copied directly.
         /// </summary>
         private static bool ToArray(BorrowedReference value, Type obType, out object? result, bool setError)
         {
             Type elementType = obType.GetElementType();
             result = null;
 
+            // structs can also be blittable but for the standard usecases this suffices
+            var isBlittable = elementType.IsPrimitive && elementType != typeof(char) && elementType != typeof(bool);
+           if (isBlittable && Runtime.PyObject_GetBuffer(value, out var view, (int)PyBUF.CONTIG_RO) == 0)
```

Comment:
I'm not sure it is safe to assume getting buffer has the same semantics as using iterator for an arbitrary type that supports GetBuffer.

Imho, this could be made the default behavior for converting a fixed set of standard types, but for the others it might be better to expose some kind of explicit call instead of conversion. E.g. `pyObjectInst.ReadBuffer(destinationArray)`

Reply:
Isn't this caught by the primitive blitable check?

Reply:
We could move the whole thing to after the `if (IterObject.IsNull()` check to make sure it's an iterable? Other than that as @filmor said I already allow only primitive blittable types as array element types.

Reply:
This issue is not about element types, but about container types. Does Python guarantee, that GetBuffer()[0], GetBuffer()[1], etc is the same sequence as .iter[0], .iter[1], etc

Reply:
I thought that was the point of the buffer protocol :). I'll read up more on it.

Reply:
I'm not sure such a guarantee can be made, wouldn't it fail for multidimensional arrays? Is it even necessary? If you define the semantics of `ToArray()` as _convert a given python object to a .NET array if possible_ than the information describing [Complex arrays](https://docs.python.org/3/c-api/buffer.html#complex-arrays) (shape, strides, format, contiguity) is I believe enough to copy such a buffer correctly to an array and to check whether the user requested a matching array in the first place. By always requesting `GetBuffer(PyBUF_FULL_RO)` we can make sure to avoid converting buffers that do not give us all information. However just doing the optimization for a certain set of types to avoid any unforseen issues may be the better option anyway.

Reply:
@chickenservice the concern is about breaking changes, especially within Python.NET 3.x, more especially within Python.NET 3.0.x.

If there's a difference in result, we should delay the change of behavior to 4.x, and in 3.x only add new functions.

IMHO, we can only immediately apply the speedups for the types that we know have the same behavior.

Reply:
@lostmsu that makes sense. The tests cover bytearray, array.array and ndarray. From my experience these behave according to what you described and are the canonical implementations anyway. Any concerns? Are there other types you have in mind? Should we limit ndim == 1 for now?

Reply:
If the tests do not cover multidim or the behavior in multidim is different, we should limit ndim to 1.

But again, I strongly suggest to add an explicit new method that does the right thing with buffers, and only works with buffer protocol.

Reply:
I think I misunderstood your suggestion, to clarify:

For upcoming minor/patch release expose this as a dedicated method and remove it from `ToArray` completely.

For the next major release reevaluate whether this can/should be done implicitly in `ToArray` and accept breaking changes.

Reply:
@chickenservice yep, that's exactly right.